fix(windows): treat powershell as not-Unix-like in Windows

akriese opened the pull request:
Fixes #21045.

This takes part of vim patch [v8.2.3071](https://github.com/vim/vim/commit/127950241e84c822d3c50f46a00d42a70d2d5cb6) to make `pwsh` and `powershell` return the same `exepath()`s (including extensions) as `cmd.exe`.

akriese committed:
fix(exepath)!: prefers extensionless for powershell

PROBLEM:
exepath("test") should prefer ".bat" when shell=powershell.
Current behavior differs from Vim 8.2.3071.

TEST CASE:
1. in a folder which is in $PATH, create files "test" "test.bat".
   - "(Get-Command test).Path test" returns "test.bat".
2. compare nvim:
     nvim --clean
     :set shell=powershell
     :echo exepath("test")
3. should returns "path\to\test.bat" (before this patch it returns "path\to\test").

SOLUTION:
After this patch, the binary files "text.exe", "test.bat", "test.com"
will be found, but the file "test" (without any extension) will not be
found (matches Vim 8.2.3071). But powershell's `where` and
`Get-Command` _do_ find the extensionless 'test' file.

But Nvim with ":set shell=cmd.exe", doesn't find "test" either (before
and after this patch), even though `where test` returns correct path in
cmd.

- `where` is a program to find files in general, not just executable files.
-`Get-Command` returning extensionless (and thus non-executable) file is
  puzzling even to Chris Dent, [PowerShell expert][1] (asked on
  Discord).

[1]: https://www.amazon.com/Mastering-PowerShell-Scripting-Automate-environment-ebook/dp/B0971MG88X

Co-authored-by: Enan Ajmain <3nan.ajmain@gmail.com>
Helped-by: erw7 <erw7.github@gmail.com>

Fixes #21045

justinmk commented:
cc @3N4N

3N4N commented on src/nvim/os/fs.c:
```diff
@@ -302,7 +302,9 @@ static bool is_executable(const char *name, char **abspath)
   FUNC_ATTR_NONNULL_ARG(1)
 {
-  const bool is_unix_shell = strstr((char *)path_tail(p_sh), "sh") != NULL;
+  const bool is_unix_shell = strstr((char *)path_tail(p_sh), "powershell") == NULL
+                             && strstr((char *)path_tail(p_sh), "pwsh") == NULL
+                             && strstr((char *)path_tail(p_sh), "sh") != NULL;
```
No need to cast into `char *` since `path_tail` already returns a char pointer.  In Vim's source, they needed to cast because `gettail`, Vim's equivalent of `path_tail`, returns a `char_u *`<sup>[1]</sup>.

```suggestion
  const bool is_unix_shell = strstr(path_tail(p_sh), "powershell") == NULL
                             && strstr(path_tail(p_sh), "pwsh") == NULL
                             && strstr(path_tail(p_sh), "sh") != NULL;
```

[1]: https://github.com/vim/vim/blob/4ae0057308c59a0dee2b452736366e49a5a98b3a/src/filepath.c#L2886-L2887

3N4N commented:
By the way, the Windows CI failure must've been a flake.  It passed fine on [my fork][1].

[1]: https://github.com/3N4N/neovim/actions/runs/3662892446/jobs/6192300808

3N4N commented:
Do we not want extensionless file to be found with `:h exepath()`?

After this patch, the binary files 'text.exe', 'test.bat', 'test.com' will be found, but the file 'test' (without any extension) will not be found.  But both `where` and `Get-Command` find the extensionless 'test' file.

But neovim with ':set shell=cmd', doesn't find 'test' either (without applying this patch), even though `where test` returns correct path in cmd.  So I don't know what should be the expected behavior.

akriese commented:
So, as I understand it, if `exepath()` is supposed to only return executables and windows does not define extension-less files as executable, this would be the desired behavior, right? Even if the result diverges from `Get-Command`...

3N4N commented:
> this would be the desired behavior, right?

Right.  As it stands now, exepath in this PR behaves correctly.

musjj commented:
Would it be a good idea if `bash.exe` (MSYS2, Git Bash, etc.) also receive the same kind of treatment?
My use-case is that I'm using `exepath` in order to find `*.bat`/`*.cmd` files, but with `shell` set to `bash.exe`, the correct path isn't returned.
Actually, why does `exepath` depend on your `shell` option in the first place? Why not just query the user's OS?

justinmk commented:
> But with `shell` set to `bash.exe`, the correct path isn't returned.
> Actually, why does `exepath` depend on your `shell` option in the first place? Why not just query the user's OS?

in Git bash on windows, files without ".exe" suffix can be executable, right? 

> After this patch, the binary files 'text.exe', 'test.bat', 'test.com' will be found, but the file 'test' (without any extension) will not be found. But both `where` and `Get-Command` find the extensionless 'test' file.
> 
> But neovim with ':set shell=cmd', doesn't find 'test' either (without applying this patch), even though `where test` returns correct path in cmd. So I don't know what should be the expected behavior.
> 
> **Edit**: `where` is a program to find files in general, not just executable files. And `Get-Command` returning extensionless (and thus non-executable) file is puzzling even to Chris Dent, [the resident PowerShell expert](https://www.amazon.com/Mastering-PowerShell-Scripting-Automate-environment-ebook/dp/B0971MG88X) (I asked him on Discord). So I say `exepath()` should _not_ find extensionless files, even if it doesn't coincide with what `where` and `Get-Command` do.

Great info, thanks! So what is the 3-line conclusion here? Please include it in the commit message.

musjj commented:
> In Git bash on windows, files without ".exe" suffix can be executable, right?

Not completely sure of the specs, but I think only script (bash, sh, etc.) executables can have no extension in Git Bash. Binary executables require `.exe` just like usual.

3N4N commented:
> > After this patch, the binary files 'text.exe', 'test.bat', 'test.com' will be found, but the file 'test' (without any extension) will not be found. But both `where` and `Get-Command` find the extensionless 'test' file.
> > But neovim with ':set shell=cmd', doesn't find 'test' either (without applying this patch), even though `where test` returns correct path in cmd. So I don't know what should be the expected behavior.
> > **Edit**: `where` is a program to find files in general, not just executable files. And `Get-Command` returning extensionless (and thus non-executable) file is puzzling even to Chris Dent, [the resident PowerShell expert](https://www.amazon.com/Mastering-PowerShell-Scripting-Automate-environment-ebook/dp/B0971MG88X) (I asked him on Discord). So I say `exepath()` should _not_ find extensionless files, even if it doesn't coincide with what `where` and `Get-Command` do.
> 
> Great info, thanks! So what is the 3-line conclusion here? Please include it in the commit message.

I'm not the author of this PR, but it should be something like this:
```
BREAKING!: exepath() does not return extentionless files in Windows with
PowerShell as shell.
```

3N4N commented:
@akriese sorry for the confusion but I didn't mean the full commit should be replaced by my suggestion.  It should've been appended, like the message of commit [`0405594`][1].

But don't change it again now.  I'm sure the maintainer will edit the commit message before merging if it needs editing.

[1]: https://github.com/neovim/neovim/commit/0405594399741babd6d935d581eac2584b289f92

justinmk commented on src/nvim/os/fs.c:
```diff
@@ -302,7 +302,9 @@ static bool is_executable(const char *name, char **abspath)
-  const bool is_unix_shell = strstr((char *)path_tail(p_sh), "sh") != NULL;
+  const bool is_unix_shell = strstr(path_tail(p_sh), "powershell") == NULL
+                             && strstr(path_tail(p_sh), "pwsh") == NULL
+                             && strstr(path_tail(p_sh), "sh") != NULL;
```
This also affects powershell in non-Windows system.

3N4N replied:
There is an `ifdef mswin` right above the function, though.

justinmk committed:
vim-patch:8.2.3061: testing the shell option is incomplete and spread out

Problem:    Testing the shell option is incomplete and spread out.
Solution:   Move shell tests to one file and increase coverage. (Yegappan
            Lakshmanan, closes vim/vim#8464)

https://github.com/vim/vim/commit/054794c20f6322bbd9482c4124041dc0a140c78e

Co-authored-by: Yegappan Lakshmanan <yegappan@yahoo.com>

justinmk committed:
vim-patch:8.2.3070: not enough testing for shell use

Problem:    Not enough testing for shell use.
Solution:   Add a bit more testing. (Yegappan Lakshmanan, closes vim/vim#8469)

https://github.com/vim/vim/commit/ffec6dd16a766180429addaa78928c773a3c9832

Co-authored-by: Yegappan Lakshmanan <yegappan@yahoo.com>

justinmk committed:
vim-patch:8.2.3071: shell options are not set properly for PowerShell

(Most of this patch is intentionally dropped, it adds far too much
special-cases for almost zero purpose: ":help shell-powershell" makes it
easy to choose powershell without spreading special-cases throughout the
codebase, randomly changing slash behavior, etc.)

Problem:    Shell options are not set properly for PowerShell.
Solution:   Use better option defaults. (Mike Willams, closes vim/vim#8459)

https://github.com/vim/vim/commit/127950241e84c822d3c50f46a00d42a70d2d5cb6

Co-authored-by: Mike Williams <mikew@globalgraphics.com>

justinmk commented:
ported vim-patch 8.2.3071 and related patches to avoid needing to revisit this in the future. 8.2.3071 is mostly N/A because all the magic special-case bullshit is not something we want.

zeertzjq commented:
Some other related patches: #19313

justinmk committed:
test: align Test_shell_options, Test_shellslash with Nvim default

'shellxquote' Nvim default was adjusted in: 131aad953c007d382cbff1d2560471b29975da87
The use of "/s" is different than Vim, and may avoid the need for `shellxquote="&|<>()@^`.

For the other shells, Nvim intentionally does not fiddle with the
various "shell*" options if 'shell' is set by the user: if the user sets
'shell', they are expected to set other "shell*" options correctly.

justinmk commented:
> We need a test case in `test/functional/vimscript/exepath_spec.lua` .

Merged to unblock, but it would still be helpful to have this test @akriese

akriese commented:
I just created this test. First time writing a test for neovim, so forgive me, if I did something wrong. Unfortunately, I don't know how to run the test on Windows with `is_os('win')` evaluating to true. I am running with `mingw32-make functionaltest` and it fails because it thinks, that I am on a not-Windows machine.

3N4N commented:
> I am running with mingw32-make functionaltest and it fails because it thinks, that I am on a not-Windows machine.

I suggest using cmake for everything.  Neovim is a cmake project; the makefile provided with it is only for convenience, and I don't think it's catered too much to Windows devs.  What I do is use something like these:

```
PS> cmake -E env DEPS_BUILD_DIR=E:/projects/neovim/deps cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=C:/apps/nvim -DCMAKE_C_COMPILER=clang -DCMAKE_C_FLAGS=-w -B .\build\debug

PS> cmake --build .\build\debug

PS> cmake -E env TEST_FILE=.\test\functional\path\to\your\test\file.lua cmake --build .\build\debug functionaltest
```

I'm not gonna explain what's happening in the commands above.  You should be able to figure out with a little bit of time.

> Edit: Also, the test for set shell=pwsh.exe will probably fail, if that version of Powershell is not installed on the system (don't know if it is installed by default)

There is a helper function in the test suite called [`set_shell_powershell`][1].  Grep and see how it's used.

[1]: https://github.com/neovim/neovim/blob/8a4285d5637c146a0ae606918a8e77063c6a5f0d/test/functional/ex_cmds/make_spec.lua#L21

3N4N commented:
Also, @akriese you should do this in a new PR.

justinmk commented:
> What I do is use something like these:
> 
> ```
> PS> cmake -E env DEPS_BUILD_DIR=E:/projects/neovim/deps cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=C:/apps/nvim -DCMAKE_C_COMPILER=clang -DCMAKE_C_FLAGS=-w -B .\build\debug
> 
> PS> cmake --build .\build\debug
> 
> PS> cmake -E env TEST_FILE=.\test\functional\path\to\your\test\file.lua cmake --build .\build\debug --target functionaltest
> ```

see also https://github.com/neovim/neovim/wiki/Building-Neovim#windows--msvc-powershell (@3N4N any improvements to make there?)

akriese commented:
> > What I do is use something like these:
> > ```
> > PS> cmake -E env DEPS_BUILD_DIR=E:/projects/neovim/deps cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=C:/apps/nvim -DCMAKE_C_COMPILER=clang -DCMAKE_C_FLAGS=-w -B .\build\debug
> > 
> > PS> cmake --build .\build\debug
> > 
> > PS> cmake -E env TEST_FILE=.\test\functional\path\to\your\test\file.lua cmake --build .\build\debug --target functionaltest
> > ```
> 
> see also https://github.com/neovim/neovim/wiki/Building-Neovim#windows--msvc-powershell (@3N4N any improvements to make there?)

I used the instructions in the README aswell as @3N4N's commands. Didn't work out of the box, maybe my environment has some issue. It eventually worked out for me when explicitly passing the `LUAJIT_LIB` path to the cmake command ([my build script](https://github.com/akriese/dotfiles/blob/main/neovim_build.ps1))
